File: src/dfs/ConvertSortedArrayToBinarySearchTree.java
package dfs;

import utility.TreeNode;

/**
 * 
 */

public class ConvertSortedArrayToBinarySearchTree
{

    public TreeNode sortedArrayToBST( int[] nums )
    {
        return buildTreeRecurse( nums, 0, nums.length - 1 );
    }

    private TreeNode buildTreeRecurse( int[] nums, int start, int end )
    {
    	if ( start > end )
    	{
    		return null;
    	}
    	
    	int rootIndex = ( end - start ) / 2 + start;
    	TreeNode root = new TreeNode( nums[rootIndex] );
    	root.left = buildTreeRecurse( nums, start, rootIndex - 1 );
    	root.right = buildTreeRecurse( nums, rootIndex + 1, end );
    	return root;
    }
}

File: Python/Tree/BinarySearchTree/ConvertSortedArrayToBinarySearchTree.py
# Definition for a binary tree node.
import unittest

# Read about enumerate in python

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class ConvertSortedArrayToBinarySearchTree(unittest.TestCase):
    def sortedArrayToBST(self, nums: list) -> TreeNode:
        result = self.sortedArrayToBSTRecursive(nums, 0, len(nums) - 1)
        return result

    def sortedArrayToBSTRecursive(self, nums: list, start: int, end: int):
        if start > end:
            return None
        elif start == end:
            return TreeNode(nums[start])
        else:
            middle = (int)((start + end) / 2)
            root = TreeNode(nums[middle])
            root.left = self.sortedArrayToBSTRecursive(nums, start, middle - 1)
            root.right = self.sortedArrayToBSTRecursive(nums, middle + 1, end)
            return root

    def test_NormalTree(self):
        sortedArray = [-10, -3, 0, 5, 9]

        self.sortedArrayToBST(sortedArray)

if __name__ == '__main__':
    unittest.main()